tasks/main.yml

```
@@ -1,25 +1,134 @@
+    owner: root
+    group: root
+    mode: '0644'
+  when: systemd_unit_file_templates is defined
+  with_items:
+    - "{{ systemd_unit_file_templates }}"
+  register: __systemd_unit_file_templates_result
+
+- name: Create drop-in directories
+  file:
+    path: "{{ __admin_units_dir }}/{{ '.'.join(item.split('.')[:-1]) }}.d/"
+    state: directory
+    owner: root
+    group: root
+    mode: '0755'
+  when: systemd_dropins is defined
+  with_items:
+    - "{{ systemd_dropins }}"
+
+- name: Deploy systemd drop in config
+  template:
+    src: "{{ item }}.j2"
+    dest: "{{ __admin_units_dir }}/\
+     {{ '.'.join(item.split('.')[:-1]) }}.d/\
+     99-override.conf"
+    owner: root
+    group: root
+    mode: '0644'
+  when: systemd_dropins is defined
+  with_items:
+    - "{{ systemd_dropins }}"
+  register: __systemd_dropins_result
+
+- name: Reload systemd
+  systemd:
+    daemon_reload: true
+  when: >
+    __systemd_unit_files_result is changed or
+    __systemd_unit_file_templates_result is changed or
+    __systemd_dropins_result is changed
+
+- name: Ensure required units are started
+  systemd:
     name: "{{ item }}"
     state: started
+  when: systemd_started_units is defined
+  with_items:
+    - "{{ systemd_started_units }}"
+
+- name: Ensure required units are stopped
+  systemd:
+    name: "{{ item }}"
+    state: stopped
+  when: systemd_stopped_units is defined
+  with_items:
+    - "{{ systemd_stopped_units }}"
+
+- name: Ensure required units are restarted
+  systemd:
+    name: "{{ item }}"
+    state: restarted
+  when: systemd_restarted_units is defined
+  with_items:
+    - "{{ systemd_restarted_units }}"
+
+- name: Ensure required units are reloaded
+  systemd:
+    name: "{{ item }}"
+    state: reloaded
+  when: systemd_reloaded_units is defined
+  with_items:
+    - "{{ systemd_reloaded_units }}"
+
+- name: Ensure required units are enabled
+  systemd:
+    name: "{{ item }}"
     enabled: true
-  loop: "{{ __template_services }}"
+  when: systemd_enabled_units is defined
+  with_items:
+    - "{{ systemd_enabled_units }}"
 
-- name: Generate /etc/{{ __template_foo_config }}
-  template:
-    src: "{{ __template_foo_config }}.j2"
-    dest: /etc/{{ __template_foo_config }}
-    backup: true
-    mode: 0400
-  notify: Handler for template to restart services
+- name: Ensure required units are disabled
+  systemd:
+    name: "{{ item }}"
+    enabled: false
+  when: systemd_disabled_units is defined
+  with_items:
+    - "{{ systemd_disabled_units }}"
+
+- name: Ensure required units are masked
+  systemd:
+    name: "{{ item }}"
+    masked: true
+  when: systemd_masked_units is defined
+  with_items:
+    - "{{ systemd_masked_units }}"
+
+- name: Ensure required units are unmasked
+  systemd:
+    name: "{{ item }}"
+    masked: false
+  when: systemd_unmasked_units is defined
+  with_items:
+    - "{{ systemd_unmasked_units }}"
+
+- name: Gather systemd units state
+  systemd_units:
```

```suggestion
  systemd_units:
  register: systemd_units
```
I think `register` puts the variable into the public namespace, so you don't need to use the `set_fact` below